### PR TITLE
Added v2 Datasource Write Support

### PR DESCRIPTION
**Pull request by rozza**
SPARK-202

### CODE REVIEW

**Comment by rozza** [comment]
Follows on from #34 and add the write support.

File: src/main/scala/com/mongodb/spark/sql/v2/DefaultSource.scala [hunk 1]
```diff
@@ -31,13 +31,34 @@ import org.bson.{BsonArray, BsonDocument, BsonType, Document}
...
+  override def createWriter(jobId: String, schema: StructType, mode: SaveMode, options: DataSourceOptions): Optional[DataSourceWriter] = {
+    val writeConfig = WriteConfig(optionsToMap(options), None)
+    val mongoConnector = MongoConnector(writeConfig.asOptions)
+    lazy val collectionExists: Boolean = mongoConnector.withDatabaseDo(
+      writeConfig, { db => db.listCollectionNames().asScala.toList.contains(writeConfig.collectionName) }
+    )
+
+    val dataSourceWriter: DataSourceWriter = MongoDataSourceWriter(writeConfig, schema, mode)
+    mode match {
+      case ErrorIfExists if collectionExists => throw new UnsupportedOperationException("MongoCollection already exists")
```
Comment (jyemin) [inline src/main/scala/com/mongodb/spark/sql/v2/DefaultSource.scala]: `MongoCollection` is a Java driver thing.  How about just "Collection exists" or "Collection already exists on MongoDB server"

File: src/main/scala/com/mongodb/spark/sql/v2/DefaultSource.scala [hunk 2]
```diff
@@ -31,13 +31,34 @@ import org.bson.{BsonArray, BsonDocument, BsonType, Document}
 
 import scala.collection.JavaConverters._
 
-class DefaultSource extends DataSourceV2 with ReadSupport with ReadSupportWithSchema {
+class DefaultSource extends DataSourceV2 with ReadSupport with ReadSupportWithSchema with WriteSupport {
```
[the remark this answers is not in the document]
Comment (jyemin) [reply src/main/scala/com/mongodb/spark/sql/v2/DefaultSource.scala]: Huzzah!

(Acknowledged)

File: src/main/scala/com/mongodb/spark/sql/v2/DefaultSource.scala [hunk 3]
```diff
@@ -31,13 +31,34 @@ import org.bson.{BsonArray, BsonDocument, BsonType, Document}
...
+    )
+
+    val dataSourceWriter: DataSourceWriter = MongoDataSourceWriter(writeConfig, schema, mode)
+    mode match {
+      case ErrorIfExists if collectionExists => throw new UnsupportedOperationException("MongoCollection already exists")
+      case Ignore if collectionExists        => Optional.empty()
+      case Overwrite =>
+        mongoConnector.withCollectionDo(writeConfig, { collection: MongoCollection[Document] => collection.drop() })
+        Optional.of(dataSourceWriter)
+      case _ => Optional.of(dataSourceWriter)
```
Comment (jyemin) [inline src/main/scala/com/mongodb/spark/sql/v2/DefaultSource.scala]: I think best practice here would be to explicitly have a case for `Append`, and throw for any unhandled enumerated values that may be added in the future.

File: src/main/scala/com/mongodb/spark/sql/v2/MongoDataSourceWriter.scala [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+      if (writeConfig.replaceDocument) SaveType.Replace else SaveType.Update
+    } else {
+      SaveType.Insert
+    }
+  }
+  val mongoConnector = MongoConnector(writeConfig.asOptions)
+
+  override def createWriterFactory(): DataWriterFactory[Row] = MongoDataWriteFactory()
+
+  override def commit(messages: Array[WriterCommitMessage]): Unit = {
```
Comment (jyemin) [inline src/main/scala/com/mongodb/spark/sql/v2/MongoDataSourceWriter.scala]: Is there an assumption in WriteSupport that the implementation supports transactions?  I'm not clear on the implications of commit and abort being no-ops

File: src/main/scala/com/mongodb/spark/sql/v2/MongoDataSourceWriter.scala [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+  }
+
+  override def abort(messages: Array[WriterCommitMessage]): Unit = {
```
Comment (jyemin) [inline src/main/scala/com/mongodb/spark/sql/v2/MongoDataSourceWriter.scala]: The Javadoc says 

>    * If this method fails (by throwing an exception), the underlying data source may require manual cleanup.

Is that not the case for MongoDB without transactions?

File: src/main/scala/com/mongodb/spark/sql/v2/MongoDataSourceWriter.scala [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+          new ReplaceOneModel[BsonDocument](queryDocument, doc, new ReplaceOptions().upsert(true))
+        } else {
+          queryDocument.keySet().asScala.foreach(doc.remove(_))
+          new UpdateOneModel[BsonDocument](queryDocument, new BsonDocument("$set", doc), new UpdateOptions().upsert(true))
+        }
+      }
+      writeOperation(op)
+    }
+
+    override def commit(): WriterCommitMessage = {
```
Comment (jyemin) [inline src/main/scala/com/mongodb/spark/sql/v2/MongoDataSourceWriter.scala]: So MongoDB can't do transactions with > 16MB even on 4.0, but it seems like we're violating the visibility guarantees documented in the Javadoc for API being implemented here.

File: src/main/scala/com/mongodb/spark/sql/v2/DefaultSource.scala (hunk 3 above)
Comment (rozza) [reply src/main/scala/com/mongodb/spark/sql/v2/DefaultSource.scala]: Done.

File: src/main/scala/com/mongodb/spark/sql/v2/MongoDataSourceWriter.scala (hunk 4 above)
Comment (rozza) [reply src/main/scala/com/mongodb/spark/sql/v2/MongoDataSourceWriter.scala]: No assumption of transactions, not many (if any?) systems can handle distributed transactions coming from multiple sources - this method just gets all write commit messages.

File: src/main/scala/com/mongodb/spark/sql/v2/MongoDataSourceWriter.scala (hunk 5 above)
Comment (rozza) [reply src/main/scala/com/mongodb/spark/sql/v2/MongoDataSourceWriter.scala]: This is an API that is going through further change.  Essentially we could potentially clean up written messages here.

File: src/main/scala/com/mongodb/spark/sql/v2/MongoDataSourceWriter.scala (hunk 6 above)
Comment (rozza) [reply src/main/scala/com/mongodb/spark/sql/v2/MongoDataSourceWriter.scala]: OK - I see its not transactional in that nature - eg the whole operation is a single transaction. So we don't need ACID guarantees of using Mongo Transactions. But it does need a refactor.

File: src/main/scala/com/mongodb/spark/sql/v2/MongoDataSourceWriter.scala (hunk 4 above)
Comment (rozza) [reply src/main/scala/com/mongodb/spark/sql/v2/MongoDataSourceWriter.scala]: To clarify I don't think this places any requirements of the underlying system being transactional - but correct this should trigger the "Spark transaction" and follow its requirements. New patch in.

**Pull request closed without merging**